isbnbot/normalize_isbns.py

```diff
@@ -0,0 +1,137 @@
+        file_handler.setLevel(logging.DEBUG)
+        file_handler.setFormatter(log_formatter)
+        self.logger.addHandler(file_handler)
+
+    @staticmethod
+    def isbn_needs_normalization(isbn: str) -> False:
+        """
+        Returns True if the given ISBN is valid and needs to be normalized (hyphens removed, letters capitalized, etc.)
+        Returns False otherwise
+        """
+        if isbnlib.is_isbn10(isbn) or isbnlib.is_isbn13(isbn):
+            normalized_isbn = isbnlib.get_canonical_isbn(isbn)  # get_canonical_isbn returns None if ISBN is invalid
+            return normalized_isbn and normalized_isbn != isbn
+        return False
+
+    def run(self, dump_filepath: str) -> None:
+        """
+        Performs ISBN normalization (removes hyphens and capitalizes letters)
+
+        dump_filepath -- path to *.txt.gz dump containing editions that need to be operated on
+        """
+        if self.dry_run:
+            self.logger.info('dry_run set to TRUE. Script will run, but no data will be modified.')
+
+        header = {'type': 0,
+                  'key': 1,
+                  'revision': 2,
+                  'last_modified': 3,
+                  'JSON': 4}
+        comment = 'normalize ISBN'
+        with gzip.open(dump_filepath, 'rb') as fin:
+            for row in fin:
+                row = row.decode().split('\t')
+                _json = json.loads(row[header['JSON']])
+                if _json['type']['key'] != '/type/edition': continue
+
+                isbns_by_type = dict()
+                if 'isbn_10' in _json:
+                    isbns_by_type['isbn_10'] = _json.get('isbn_10', None)
+                if 'isbn_13' in _json:
+                    isbns_by_type['isbn_13'] = _json.get('isbn_13', None)
+                if not isbns_by_type: continue
+
+                skip_flag = True
+                for _, isbns in isbns_by_type.items():
+                    for isbn in isbns:
+                        if self.isbn_needs_normalization(isbn):
+                            skip_flag = False
+                            break
+                if skip_flag: continue
+
+                olid = _json['key'].split('/')[-1]
+                edition = self.ol.Edition.get(olid)
+                if edition.type['key'] != '/type/edition': continue
+
+                for isbn_type, isbns in isbns_by_type.items():
+                    isbns = getattr(edition, isbn_type, [])
+                    for isbn in isbns:
+                        if not self.isbn_needs_normalization(isbn): continue
+                        normalized_isbn = isbnlib.get_canonical_isbn(isbn)
+                        setattr(edition, isbn_type, [normalized_isbn])  # FIXME, what if an edition has multiple isbns that need normalization?
+                        self.logger.info('\t'.join([olid, isbn, normalized_isbn]))
+                        self.save(lambda: edition.save(comment=comment))
+
+    def save(self, save_fn):
```

Comment:
Actually, it would be good if we could specify the edition key (or something) in the `dry_run` case; maybe let's make this signature like so:

```suggestion
    def save(self, record, comment):
```

Reply:
Wouldn't line 103 accomplish this for all cases?

Reply:
:+1: Ah, missed that; yep! Looks good.